core/trino-main/src/main/java/io/trino/sql/planner/LocalDynamicFilterConsumer.java

```diff
@@ -122,33 +132,41 @@ public void setPartitionCount(int partitionCount)
             if (collected) {
                 return;
             }
-            checkState(expectedPartitionCount == PARTITION_COUNT_INITIAL_VALUE, "setPartitionCount should be called only once");
-            requireNonNull(partitions, "partitions is null");
+            checkState(expectedPartitionCount == null, "setPartitionCount should be called only once");
             expectedPartitionCount = partitionCount;
-            if (partitions.size() == expectedPartitionCount) {
-                // No more partitions are left to be processed.
-                if (partitions.isEmpty()) {
-                    result = TupleDomain.none();
-                }
-                else {
-                    result = TupleDomain.columnWiseUnion(partitions);
-                }
-                collected = true;
-                partitions = null;
+            if (collectedPartitionCount < expectedPartitionCount) {
+                return;
+            }
+            if (partitionCount == 0) {
+                result = TupleDomain.all();
             }
             else {
-                return;
+                // run final compaction as previous concurrent compactions may have left more than a single domain
+                compactSummaryDomains();
+                verify(summaryDomains.size() == 1);
+                result = summaryDomains.poll();
+                verify(result != null);
             }
+            collected = true;
         }
 
-        notifyConsumers(result);
+        collectors.forEach(collector -> collector.accept(convertTupleDomain(result)));
     }
 
-    private void notifyConsumers(TupleDomain<DynamicFilterId> result)
+    private void compactSummaryDomains()
```

Comment:
In this implementation the compaction is happening outside the lock and will not block other thread from completion. By looking at the union implementation the total amount of work should not be increased insignificantly (value sets are merged 2 at a time, it will only create more small objects along the way).